### PR TITLE
proposed change for common.states

### PR DESCRIPTION
**Pull request by Apollon77**

### CODE REVIEW

**Review by foxriver76 — APPROVED** [review]

File: docs/en/dev/objectsschema.md [hunk 1]
```diff
@@ -200,7 +200,12 @@ Attributes:
 * `common.read`   (boolean, mandatory) - true if state is readable
 * `common.write`  (boolean, mandatory) - true if state is writable
 * `common.role`   (string,  mandatory) - role of the state (used in user interfaces to indicate which widget to choose, see below)
-* `common.states` (optional) attribute of type number with the object of possible states `{'value': 'valueName', 'value2': 'valueName2', 0: 'OFF', 1: 'ON'}` or (supported up from admin5) an states array, like `['Start', 'Flight', 'Land']`
+* `common.states` (optional) - provide more context on allowed values for states with datatypes string and number:
+  * for numbers **without** provided common.min/common.max: contains the list of allowed number values and their textual mapping as object in form `{0: 'OFF', 1: 'ON', '-1': 'whatever'}`. Only these values should be allowed
```
Comment (AlCalzone) [inline docs/en/dev/objectsschema.md:204]: **Should be** or **are**?

File: docs/en/dev/objectsschema.md [hunk 2]
```diff
@@ -200,7 +200,12 @@ Attributes:
 * `common.read`   (boolean, mandatory) - true if state is readable
 * `common.write`  (boolean, mandatory) - true if state is writable
 * `common.role`   (string,  mandatory) - role of the state (used in user interfaces to indicate which widget to choose, see below)
-* `common.states` (optional) attribute of type number with the object of possible states `{'value': 'valueName', 'value2': 'valueName2', 0: 'OFF', 1: 'ON'}` or (supported up from admin5) an states array, like `['Start', 'Flight', 'Land']`
+* `common.states` (optional) - provide more context on allowed values for states with datatypes string and number:
+  * for numbers **without** provided common.min/common.max: contains the list of allowed number values and their textual mapping as object in form `{0: 'OFF', 1: 'ON', '-1': 'whatever'}`. Only these values should be allowed
+  * for numbers **with** provided common.min **and/or** common.max: the allowed number range is definedby min/max, this attribute contains a list of "special" number values and their textual mapping as object like `{0: 'OFF', 254: 'ON', 255: 'BLINK'}` (min=0, max=255).
```
Comment (AlCalzone) [inline docs/en/dev/objectsschema.md:205]: What's the range if only min or max are given?

File: docs/en/dev/objectsschema.md [hunk 3]
```diff
@@ -200,7 +200,12 @@ Attributes:
...
 * `common.role`   (string,  mandatory) - role of the state (used in user interfaces to indicate which widget to choose, see below)
-* `common.states` (optional) attribute of type number with the object of possible states `{'value': 'valueName', 'value2': 'valueName2', 0: 'OFF', 1: 'ON'}` or (supported up from admin5) an states array, like `['Start', 'Flight', 'Land']`
+* `common.states` (optional) - provide more context on allowed values for states with datatypes string and number:
+  * for numbers **without** provided common.min/common.max: contains the list of allowed number values and their textual mapping as object in form `{0: 'OFF', 1: 'ON', '-1': 'whatever'}`. Only these values should be allowed
+  * for numbers **with** provided common.min **and/or** common.max: the allowed number range is definedby min/max, this attribute contains a list of "special" number values and their textual mapping as object like `{0: 'OFF', 254: 'ON', 255: 'BLINK'}` (min=0, max=255).
+  * for strings contains the list of allowed values and their textial mapping as object like `{'value': 'valueName', 'value2': 'valueName2'}`. Only these values should be allowed
```
Comment (AlCalzone) [inline docs/en/dev/objectsschema.md:206]: textial mapping -> (displayed) label

File: docs/en/dev/objectsschema.md [hunk 4]
```diff
@@ -200,7 +200,12 @@ Attributes:
...
+* `common.states` (optional) - provide more context on allowed values for states with datatypes string and number:
+  * for numbers **without** provided common.min/common.max: contains the list of allowed number values and their textual mapping as object in form `{0: 'OFF', 1: 'ON', '-1': 'whatever'}`. Only these values should be allowed
+  * for numbers **with** provided common.min **and/or** common.max: the allowed number range is definedby min/max, this attribute contains a list of "special" number values and their textual mapping as object like `{0: 'OFF', 254: 'ON', 255: 'BLINK'}` (min=0, max=255).
+  * for strings contains the list of allowed values and their textial mapping as object like `{'value': 'valueName', 'value2': 'valueName2'}`. Only these values should be allowed
+  * for strings contains the list of allowed values as array like `['Start', 'Flight', 'Land']` (which in fact is the same as `{'Start': 'Start', 'Flight': 'Flight', 'Land': 'Land'}`)
```
Comment (AlCalzone) [inline docs/en/dev/objectsschema.md:207]: Maybe add that Only these values are allowed?

**Review by AlCalzone — CHANGES_REQUESTED** [review]

File: docs/en/dev/objectsschema.md (hunk 1 above)
Comment (Apollon77) [reply docs/en/dev/objectsschema.md:204]: Thats the question how strict we want to define it. in my eyes "are" :-) In fact the question how we check it in the end

File: docs/en/dev/objectsschema.md (hunk 2 above)
Comment (Apollon77) [reply docs/en/dev/objectsschema.md:205]: As now too ... 
* min <= x < "Infinity" or 
* -Infinity > x >= max

File: docs/en/dev/objectsschema.md (hunk 3 above)
Comment (Apollon77) [reply docs/en/dev/objectsschema.md:206]: good idea. I will adjust it

File: docs/en/dev/objectsschema.md (hunk 4 above)
Comment (Apollon77) [reply docs/en/dev/objectsschema.md:207]: kk

File: docs/en/dev/objectsschema.md (hunk 2 above)
Comment (Apollon77) [reply docs/en/dev/objectsschema.md:205]: DO we need to write that down or is this common sense?

Comment (Apollon77) [reply docs/en/dev/objectsschema.md:205]: I added text

**Review by AlCalzone — APPROVED** [review]

**Comment by Apollon77** [comment]
> definedby => defined by

fixed

**Comment by Apollon77** [comment]
For transparency: we decided against a "array representation for number values" because:
* it might not be clear that array index would be the number - and a "un-thought change" by inserting a new value would change number meanings
* It might get confused with the "meaning" for strings becaus looks the same

In the end: it would be risky

**Comment by Apollon77** [comment]
I created issues for the "non Bluefox visu adapters" to give sopme days for feedback, lets see if we get reactions there. I will merge in some days.

Then list of visus that should (could maybe) adjusted if we like:

* vis - Bluefox
* (admin) - Bluefox
* jarvis - zefau
* habpanel - Bluefox?
* igopro - nis
* iqontrol - SBohrmann
* lovelace - Garfonso/Bluefox
* material - Bluefox
* minuvis - svallant
* (mobile) - Bluefox
* tileboard - Bluefox
* echarts - Bluefox
* eventlist? - Bluefox
* (flot) - Bluefox

**Comment by Garfonso** [comment]
Hi, 

as feedback was requested in the [issue](https://github.com/ioBroker/ioBroker.lovelace/issues/359): 
Lovelace already supports the String-Array and String-String-Records (and of course Number-String Record). So I'm heavily in favor for this change. :smiley:  Especially String-Array is much easier for people to understand for string-states with only a restrcited set of options. 

Since I do not have full control of the UI, lovelace can probably not (or only in some cases) support the Number-String Record with min/max. But I think they still make sense. Let's see how they get used.

Thanks for this proposal.

**Pull request merged**